py/server/deephaven/arrow.py

```diff
@@ -0,0 +1,148 @@
+#
+#     Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+"""This module supports conversions between Arrow tables and Deephaven tables."""
+
+from typing import List, Dict
+
+import jpy
+import pyarrow as pa
+
+from deephaven import DHError
+from deephaven.table import Table
+
+_JArrowToTableConverter = jpy.get_type("io.deephaven.extensions.barrage.util.ArrowToTableConverter")
+_JTableToArrowConverter = jpy.get_type("io.deephaven.extensions.barrage.util.TableToArrowConverter")
+
+_ARROW_DH_DATA_TYPE_MAPPING = {
+    pa.null(): '',
+    pa.bool_(): 'java.lang.Boolean',
+    pa.int8(): 'byte',
+    pa.int16(): 'short',
+    pa.int32(): 'int',
+    pa.int64(): 'long',
+    pa.uint8(): '',
+    pa.uint16(): 'char',
+    pa.uint32(): '',
+    pa.uint64(): '',
+    pa.float16(): '',
+    pa.float32(): 'float',
+    pa.float64(): 'double',
+    pa.time32('s'): '',
+    pa.time32('ms'): '',
+    pa.time64('us'): '',
+    pa.time64('ns'): 'io.deephaven.time.DateTime',
+    pa.timestamp('s', tz=None): '',
+    pa.timestamp('ms', tz=None): '',
+    pa.timestamp('us', tz=None): '',
+    pa.timestamp('ns', tz=None): '',
+    pa.date32(): 'java.time.LocalDate',
+    pa.date64(): 'java.time.LocalDate',
+    pa.duration('s'): '',
+    pa.duration('ms'): '',
+    pa.duration('us'): '',
+    pa.duration('ns'): '',
+    pa.month_day_nano_interval(): '',
+    pa.binary(): '',
+    pa.string(): 'java.lang.String',
+    pa.utf8(): 'java.lang.String',
+    pa.large_binary(): '',
+    pa.large_string(): '',
+    pa.large_utf8(): '',
+    # decimal128(int precision, int scale=0)
+    # list_(value_type, int list_size=-1)
+    # large_list(value_type)
+    # map_(key_type, item_type[, keys_sorted])
+    # struct(fields)
+    # dictionary(index_type, value_type, …)
+}
+
+SUPPORTED_ARROW_TYPES = [k for k, v in _ARROW_DH_DATA_TYPE_MAPPING.items() if v]
+
+
+def _map_arrow_type(arrow_type) -> Dict[str, str]:
+    """Maps an Arrow type to the corresponding Deephaven column data type."""
+    dh_type = _ARROW_DH_DATA_TYPE_MAPPING.get(arrow_type)
+    if not dh_type:
+        # if this is a case of timestamp with tz specified
+        if isinstance(arrow_type, pa.TimestampType) and arrow_type.tz:
+            dh_type = "io.deephaven.time.DateTime"
+
+    if not dh_type:
+        raise DHError(message=f'unsupported arrow data type : {arrow_type}, refer to '
+                              f'deephaven.arrow.SUPPORTED_ARROW_TYPES for the list of supported Arrow types.')
+
+    return {"deephaven:type": dh_type}
```

Comment:
1) Together, these two are inconsistent.  We do support timestamps if they have timezones, but that is not reflected in `SUPPORTED_ARROW_TYPES`.

2) Are we not supporting cases without timezones to force users to specify the timezone or for some other reason?

3) I think that no specified timezone is traditionally interpreted as the machine local timezone.  https://arrow.apache.org/docs/python/timestamps.html

Reply:
1. as one of the skipped test cases shows, the Arrow->DH conversion of timestamp isn't done correctly. The problem is tracked by #3223. So the SUPPORTED_ARROW_TYPES is correct even though the issue with 'timestamp' isn't strictly 'not supported'.

Reply:
I have concrerns that `timestamps` return values, but the values are wrong.  Users are not very tolerant of corrupt data.